Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,258 @@
+        {
+            Logger.Debug(String.Format("Loading Credential from file {0}", credentialPath));
+
+            using (Stream stream = GetStream(credentialPath))
+            {
+                return CreateDefaultCredentialFromStream(stream);
+            }
+        }
+
+        /// <summary>Creates default credential from a stream that contains JSON credential data.</summary>
+        internal GoogleCredential CreateDefaultCredentialFromStream(Stream stream)
+        {
+            JsonCredentialParameters credentialParameters;
+            try
+            {
+                credentialParameters = NewtonsoftJsonSerializer.Instance.Deserialize<JsonCredentialParameters>(stream);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException("Error deserializing JSON credential data.", e);
+            }
+            return CreateDefaultCredentialFromJson(credentialParameters);
+        }
+
+        /// <summary>Creates a default credential from JSON data. </summary>
+        private static GoogleCredential CreateDefaultCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            switch (credentialParameters.Type)
+            {
+                case JsonCredentialParameters.AuthorizedUserCredentialType:
+                    return GoogleCredential.FromCredential(CreateUserCredentialFromJson(credentialParameters));
+
+                case JsonCredentialParameters.ServiceAccountCredentialType:
+                    return GoogleCredential.FromCredential(CreateServiceAccountCredentialFromJson(credentialParameters));
+                
+                default:
+                    throw new InvalidOperationException(String.Format("Error creating credential from JSON. Unrecognized credential type {0}.", credentialParameters.Type));
+            }
+        }
+
+        private static UserCredential CreateUserCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            if (credentialParameters.Type != JsonCredentialParameters.AuthorizedUserCredentialType ||
+                string.IsNullOrEmpty(credentialParameters.ClientId) ||
+                string.IsNullOrEmpty(credentialParameters.ClientSecret))
+            {
+                throw new InvalidOperationException("JSON data do not represent a valid user credential.");
+            }
+
+            var initializer = new GoogleAuthorizationCodeFlow.Initializer
+            {
+                ClientSecrets = new ClientSecrets
+                {
+                    ClientId = credentialParameters.ClientId,
+                    ClientSecret = credentialParameters.ClientSecret
+                }
+            };
+
+            var token = new TokenResponse
+            {
+                RefreshToken = credentialParameters.RefreshToken
+            };
+
+            var flow = new GoogleAuthorizationCodeFlow(initializer);
```

Comment:
I would change the order of the code.
var token... should be before the initialzier.
initializer should be close to the place that it's being used...